pkg/apiserver/certificate/certificate.go

```diff
@@ -124,15 +127,14 @@ func generateSelfSignedCertificate(secureServing *options.SecureServingOptionsWi
 	var caContentProvider dynamiccertificates.CAContentProvider
 
 	// Set the PairName but leave certificate directory blank to generate in-memory by default.
-	secureServing.ServerCert.CertDirectory = ""
+	secureServing.ServerCert.CertDirectory = selfSignedCertDir
 	secureServing.ServerCert.PairName = "antrea-controller"
 
 	if err := secureServing.MaybeDefaultWithSelfSignedCerts("antrea", GetAntreaServerNames(), []net.IP{net.ParseIP("127.0.0.1")}); err != nil {
 		return nil, fmt.Errorf("error creating self-signed certificates: %v", err)
 	}
 
-	certPEMBlock, _ := secureServing.ServerCert.GeneratedCert.CurrentCertKeyContent()
-	caContentProvider, err = dynamiccertificates.NewStaticCAContent("self-signed cert", certPEMBlock)
+	caContentProvider, err = dynamiccertificates.NewDynamicCAContentFromFile("self-signed cert", secureServing.ServerCert.CertKey.CertFile)
```

Comment:
@tnqn I was reading the code and I realized that in the self-signed case, we actually do not use a self-signed CA, we just self-sign a certificate directly and distribute this as the "antrea-ca" ConfigMap. Was that the intention? I seem to recall that initially you wanted to use a self-signed CA, and use that to sign a certificate for the Controller apiserver.

Unless I am mistaken about the above, I find some terminology to be confusing, along with our documentation:
> By default, antrea-controller generates a self-signed certificate. You can override the behavior by providing your own certificates. Either way, the antrea-controller will distribute the CA certificate as a ConfigMap named antrea-ca in the Antrea deployment Namespace and inject it into the APIServices resources created by Antrea in order to allow its clients (i.e. antrea-agent, kube-apiserver) to perform authentication.

> Typically, clients that wish to access the antrea-controller API can authenticate the server by validating against the CA certificate published in the antrea-ca ConfigMap.

Reply:
The function `GenerateSelfSignedCertKeyWithFixtures` does [create a self-signed CA certificate](https://github.com/kubernetes/kubernetes/blob/e7de96e635f2dc50b2c5427743473093f4486d28/staging/src/k8s.io/client-go/util/cert/cert.go#L131) and [use it to sign a certificate for antrea-controller](https://github.com/kubernetes/kubernetes/blob/e7de96e635f2dc50b2c5427743473093f4486d28/staging/src/k8s.io/client-go/util/cert/cert.go#L168), and only the second cert includes the required DNS names.

The only thing that may be confusing is both the CA certificate and the TLS certificate are returned by the function and published to antrea-ca configmap, but both of them can be used to verify the server. If we want to strictly publish the CA certificate only, we could extract it from the return value.

Reply:
Thanks Quan for the explanation. I was indeed confused by the fact that they are "bundled" together, but that makes sense now.